Fix the type of global not carrying across files

JanSharp opened the pull request:
This fixes
```lua
---@class MyGlobal
global = {}
```
not affecting the type of global in any file outside of the one with this annotation. This seems to be caused by the language server having some understanding of lifetimes of globals, as doing `global = {}` makes it lose type information in code folloing that statement, or something.

The fix works by removing the plugin injected assignment to global at the top of each file.
Instead it adds a diagnostic disable on each usage of the renamed global.

There are newlines before and after this diagnostic disable as to not affect any other code.

These newlines do not cause any other issues themselves, however in order to position the preceeding one properly it had to include the preceeding character before the word `global` in the diff. This can cause said character to get duplicataed if there is another diff that also includes this character, such as code like `remote.call("foo", "bar")global.foo = "bar"`. Note how there is no space between `)` and `global`. I consider this to be small enough of an issue to be worth getting undefined global warnings in cases like `foo.bar = global.bar` for `foo`.

The newline after the injected diagnostic disable does not have any side effects that I'm aware of, so this gives a warning that the global `foo` does not exist `global.bar = foo`.

Overall this does worsen the performance of the plugin, however not by much (a more complex pattern for globals, including creating strings for all preceeding text on all lines containing `global`, similar to the logic for `remote`). If it's a concern I can do measurements, however I'm not worried.

JanSharp committed:
Fix the type of global not carrying across files

JanSharp commented:
Wait, hold up. If it's adding a newline before and after anyway, there's no need for the block comment. I gotta fix that real quick.

JanSharp committed:
Remove unnecessary usage of block comments

Therefore also improve performance

justarandomgeek commented:
![image](https://github.com/justarandomgeek/vscode-factoriomod-debug/assets/5246445/948099d9-5b82-4ef2-9f13-2c7dc3d4f888)
this combined with the remote.call rewrite seems to cause some trouble

```lua
remote.call("conman","set_preload_string",global.conman.unit_number,"TEST")
```
produces these errors, but adding a space before `global` fixes it: 

```lua
remote.call("conman","set_preload_string", global.conman.unit_number,"TEST")
```


If it matters, this is in a mod scenario control.lua (as shown in the file path in the screenshot)

JanSharp commented:
Damn it, I guess it really can't just blindly assume that it won't cause issues, because clearly it did. There is a fairly optimized way to check for an existing diff at that location, and I think it's worth adding that in order to still get undefined global warnings on the same line as (and before) `global`. I'll implement it and try to do some crude performance measuremnts.

justarandomgeek commented on luals-addon/factorio/factorio-plugin/global.lua:
```diff
@@ -52,33 +52,39 @@ local function replace(uri, text, diffs)
+        matches_to_ignore[start] = true
       end
     end
-    -- `_ENV.global` and `_G.global` now get removed because of this, we can add them back in
```
unrelated to the current issue: couldn't you look just before the dotted match just found and skipped for the parent instead of re-doing the whole search, to get _ENV and _G back included?

JanSharp replied:
Oh yea absolutely, good idea. I'll just include that in this PR even though it's unrelated

JanSharp replied:
It just won't be able to catch them if there's a space after `_G` or `_ENV` (which for some reason the commented out code doesn't check for either), but I think that's still fine, considering 99% of the time there won't be a space.

JanSharp committed:
Include usage of global even when using _G or _ENV

JanSharp committed:
Fix duplicated diffs before `global`

JanSharp committed:
Remove file_id logic and clear table every time

Turns out this is a little faster than the file_id + full reset every
  100 file passes logic, even on a small file with just a few diffs.
  Having to reallocate a big table every 100 passes for big files is
  bad, and this approach now doesn't do that, the table will get as big
  as big as it needs to be for the biggest file and stays that way,
  without further allocations.
Performance testing:
  before deduplication: 9271.325125ms
	with deduplication, resetting every 100 passes: 9383.937886ms
	with deduplication, clearing the table every time: 9356.621231ms

JanSharp commented:
Fixed, and it even emits the warning on the left side :D
![image](https://github.com/justarandomgeek/vscode-factoriomod-debug/assets/58273964/5bf50f04-2f78-46e7-85ef-44a04a10f536)
To my surprise performace is barely impacted, but I suppose that's string operations vs table operations. The newly added logic contains very few allocations, so I guess that makes sense.
